### PR TITLE
update workflow and project structure of devGuide

### LINKED ISSUE #1012: Dev Guide: Add documentation for currently used markdown-it plugins and rules
**Is your request related to a problem?**

Currently, we are using quite a significant number of markdown-it plugins and rules for current features, and will be using more plugins and rules for future features (especially for fenced code blocks). It would be difficult to keep track of which plugins are used and for which functionality. 

**Describe the solution you'd like**

Documentation of the plugins would be nice to have. Just a brief overview of
- What plugins/rules we have
- Which plugins/rule enable which functionality.

**Describe alternatives you've considered**

Currently I have to search through the past and current issues and PRs to get an overview. So this is simply a nice-to-have feature.

### LINKED ISSUE #1707: Creating vs forking vs duplicating markdown-it plugins
Hi guys,

I am working on a version update for the markdown-it package and am curious what were the thoughts behind our approach in maintaining markdown-it packages. At the moment, it seems like there are three types of markdown-it plugins:
1. plugins that we use as-is, and shown in package.json
  - e.g. markdown-it-task-lists
2. plugins that we have modified/patched slightly, also shown in package.json:
  - e.g. markdown-it-attrs
3. plugins that we copy the source code and keep in core/src/lib/plugins, not in package.json 
  - e.g. markdown-it-center-text

Not too sure if there are any discussions on this before, but curious what were the considerations when deciding which approach to go for (maybe also in general)?

Some questions I have:
- should we fork the original plugin and maintain it as a forked plugin?
  - especially when the original plugin (or packages like live-server) is no longer updated in a few years.
- should we create individual markdown-it plugins for the new functionalities that we introduced?
  - especially if the plugin has a general purpose that could potentially benefit others
- when should we decide to copy the source code and patch the code within our repo?
  - the potential pitfall from this is that version updates may become more troublesome since our patch is targeted towards a specific version of the package

Any discussions or thoughts on this would be appreciated! 
p.s If this is a useful topic, perhaps I can help write it up in our documentation with some guidelines so that we can have a standardized approach to package management and allow future devs to know what to do when faced with such a choice:)

### PR DESCRIPTION
**Pull request by tlylt**
**What is the purpose of this pull request?**

- [x] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

**Overview of changes:**
Fixes #1707
- Add a tooltip that includes a link to Babel's readme on the pros and cons of a monorepo
- Update the project structure section to include some more details about the external packages, pointing out those that have been patched
- Update some points for consideration when adding dependencies, under workflow

Potentially fixes #1012
 - The issue asks for a list of markdown-it plugins & rules to be included in the DG. I have added some descriptions about existing markdown-it plugins and also refer to the `package.json` for a full list. I feel that this might just be enough without providing too many details as to which exact rules/plugins are used, which can be subjected to (somewhat frequent) changes across versions. However, if there is still merit, I shall remove this part.

**Anything you'd like to highlight / discuss:**
- The above point on whether this PR is sufficient to close the related issue.
- Any proofreading, thoughts on the added content is appreciated.

**Testing instructions:**
See the two pages in the dev guide that have been updated
- [workflow](https://deploy-preview-1786--markbind-master.netlify.app/devguide/workflow#points-for-consideration)
- [project structure](https://deploy-preview-1786--markbind-master.netlify.app/devguide/design/projectstructure)

**Proposed commit message: (wrap lines at 72 characters)**
Update workflow and project structure of devGuide

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

**Review by ryoarmanda — CHANGES_REQUESTED** [review]
Overall looks great! Some small things that we need to address first and we're all set :)

What you've done for the Points of Consideration part is good. About whether we include a list of our patches, I'm kind of on the fence about it and can be discussed with others (cc @jonahtanjz any thoughts?)

Most of the patches I think are quite documented in their respective files e.g. `live-server`, `nunjucks` (so devs can read the top-level comment for context), but having a list of it seems fine too to get an overall picture on what we have patched so far in the repo but therefore has to be continually updated whenever we add/remove something (though I don't think we have so many of it right now, save for the `markdown-it` stuff).

File: docs/devGuide/workflow.md [hunk 1]
```diff
@@ -242,3 +242,19 @@ First, follow the instruction to [delete the dependency](#to-delete-a-dependency
...
+Installing | <ul><li>Ease of upgrade</li><li>Traceable in `package.json`</li></ul>                                                                                                | <ul><li>May not satisfy custom behavior</li><li>Become vulnerable if the source repo is no longer actively maintained</li></ul>
+Forking           | <ul><li>Relatively easy to upgrade</li><li>Leverage upstream testing procedures</li><li>Benefit others who share the same use cases</li></ul>                                                           | <ul><li>May become out-of-sync with the latest version</li><li>Difficult to maintain it externally in the long run</li></ul>                   
+Patching          | <ul><li>Quick - no need to maintain / publish more npm packages or setup release procedures etc</li><li>Ensure the changes propagate to other dependencies</li><li>Benefits of a monorepo</li></ul> | <ul><li>Difficult to upgrade</li></ul>
```
Comment (ryoarmanda) [inline docs/devGuide/workflow.md:253]: Maybe paraphrase the `Benefits of a monorepo` part to `Enjoy the benefits of monorepo`?

File: docs/devGuide/design/projectStructure.md [hunk 2]
```diff
@@ -32,29 +32,30 @@ The MarkBind project is developed in a monorepo ([MarkBind/markbind](https://git
...
+* Unit Tests (though there are more unit tests and functional tests in the cli library)
 
 **The key external libraries used are:**
 
 * [markdown-it](https://github.com/markdown-it/markdown-it), which does the Markdown parsing and rendering. There are also several customized markdown-it plugins used in MarkBind, which are located inside the `src/lib/markdown-it/` directory.
-
+  * Serveral markdown-it plugins are installed to enhance the existing Markdown syntax. They can be found in `src/package.json`. Some of them are patched in the `src/lib/markdown-it/patches/` directory to fit MarkBind's needs. 
+  * Additionally, there are some markdown-it plugins in the `src/lib/markdown-it/plugins/` directory (either forked, modified or written to enhance existing functionalities).
 * [htmlparser2](https://github.com/fb55/htmlparser2), a speedy and forgiving html parser which exposes a dom-like object structure to work on. To comply with the markdown spec, and our custom requirements, `src/patches/htmlparser2.js` patches various behaviours of this library.
-
```
[the remark this answers is not in the document]
Comment (tlylt) [reply docs/devGuide/design/projectStructure.md:50]: Hi @ryoarmanda, I tried adding blank lines, but it seems to break the interpretation of a list and resulted in weird margins

**Comment by kaixin-hc** [comment]
It looks good to me as well!

I think maintaining an active list isn't a good idea because of the maintaining will be difficult, and highlighting the relevant files as you have is a better choice.

I feel like [zeyu's extended explanation in the issue](https://github.com/MarkBind/markbind/issues/1707#issuecomment-1013679201) on why particular patches were patches instead of forks was really helpful though for giving context as to why these questions were questions that should be considered. Maybe some of that can be incorporated?

File: docs/devGuide/design/projectStructure.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryoarmanda) [reply docs/devGuide/design/projectStructure.md:50]: Hmm I see, this looks like a styling quirk, as you can do the same thing here in GitHub and get an evenly spaced list like below (see the raw if you would like).

> * First item
>   * First sub-item
>   * Second sub-item
>
> * Second item
>
> * Third item

The way to work around it right now is to add another blank line between the sub-items, which will give you the evenly spaced list as intended, but with a side effect of the sub-items being spaced out as well. You can try it out first, and see if it is still stylistically good. If not, then we can settle with no blank lines for now.

[the remark this answers is not in the document]
Comment (tlylt) [reply docs/devGuide/design/projectStructure.md:50]: Thanks for tip, fixed.

**Comment by jonahtanjz** [comment]
> What you've done for the Points of Consideration part is good. About whether we include a list of our patches, I'm kind of on the fence about it and can be discussed with others (cc @jonahtanjz any thoughts?)

Agree with the rest, maintaining a full list of patches can be easy to miss out when they get updated, which can confuse some new devs when they are looking through. I think having a list of external libraries that we have patched/fork is sufficient in the workflow page. Perhaps can also include a link under "Points for consideration" to the list of external libraries in the workflow page?

**Comment by tlylt** [comment]
Thanks for all the feedback given!
I have:
- fixed the list spacing issues in the project structure page
- put in a link from workflow to project structure
- edited some minor things based on the suggestions

> I feel like https://github.com/MarkBind/markbind/issues/1707#issuecomment-1013679201 on why particular patches were patches instead of forks was really helpful

I agree with this but am not too sure how to include it succinctly. Given that I have pointed out all the important external libraries used in the project structure, I am guessing this might be enough for readers to explore the relevant PRs/Issues. Also, I mentioned that
> Also, the rationales behind most existing patches are documented in their respective files, read them (and their respective PRs/issues) for more context!

I hope this is enough, but let me know if it isn't.

> maintaining a full list of patches can be easy to miss out when they get updated

I share the same feeling here, which is why I would like this PR to close #1012. I have mentioned the places to look for markdown-it related plugins and rules so readers can now locate them easily.

Please feel free to let me know if there are any further changes required.

**Review by ryoarmanda — APPROVED** [review]
Thanks for the further work! I think it looks good 👍

**Pull request merged**